### DISCUSSION TITLE
Where do the extra DDPM sampling steps come from?

### DISCUSSION THREAD

**Discussion by AlejandroBaron** [huggingface/diffusers#9431] [Q&A]
I see this line matching the original paper

https://github.com/huggingface/diffusers/blob/1e8cf2763d37fa66cfba2fb87acfe3443068ce43/src/diffusers/schedulers/scheduling_ddpm.py#L449

But some extra steps are in place

https://github.com/huggingface/diffusers/blob/1e8cf2763d37fa66cfba2fb87acfe3443068ce43/src/diffusers/schedulers/scheduling_ddpm.py#L468-L476

Sampling Algorithm 2 does not perform this operation in the [original paper](https://arxiv.org/pdf/2006.11239)
![image](https://github.com/user-attachments/assets/ce07bbb7-5a9d-44e6-9f0e-18bf42963317)

Where do these come from?

**Comment by bmaxdk** [comment]
These additional steps defines the calculation of the predicted sample during the reverse diffusion process which introducing coefficients that balance the contributions of the predicted original sample x_0 and the current noisy sample x_t. This will be improving both numerical stability and sample quality in the final output. That's true that those are not shown in the original paper but have likely been added based on practical considerations and optimization for real-world use cases

**Comment by jadevaibhav** [comment]
These formula comes from reverse diffusion process, where at each time step we sample from a Gaussian distribution (Step 4). The first term is basically predicted mean (computed using current sample and predicted noise). The implementation uses reformulation of the same.
 For 3 different models of training , namely prediction of 'noise', 'sample' and 'velocity prediction'; its easier to use reformulation with predicted original sample(pred_x_0) and current sample (x_t). I think that's what the code does.
